xdsl/traits.py

```diff
@@ -30,3 +30,18 @@ def verify(self, op: Operation) -> None:
             )
         names = ", ".join([f"'{p.name}'" for p in self.parameters])
         raise VerifyException(f"'{op.name}' expects parent op to be one of {names}")
+
+
+class NoTerminator(OpTrait):
+    """
+    Allow an operation to have single block regions with no terminator.
+
+    https://mlir.llvm.org/docs/Traits/#terminator
+    """
+
+    def verify(self, op: Operation) -> None:
+        for r in op.regions:
+            if len(r.blocks) > 1:
```

Comment:
It is actually a bit more complex than that. Here is the "documentation" I could find in MLIR:
![image](https://github.com/xdslproject/xdsl/assets/20801162/0a9a2754-6dde-427c-8a02-cbdac6f8278a)

In our case, we don't actually need to verify anything in the trait. What we need is somewhere a check that blocks have terminator if the conditions are respected. I'm just not sure where to put it yet, I'll take a look now.

Reply:
yeah, I found that as well yesterday, good to see, but when I tried to check for the `NoTerminator` trait in `ir.py` (simulating this), I got a circular dependency in my imports, so that kinda put me off from going that route.

Moreover, I couldn't find any validation for terminators in xDSL.

> Looking at the MLIR codebase, the check that blocks have terminators should be done in the `verify` function in `Operation` between these two code blocks: ![image](https://user-images.githubusercontent.com/20801162/242616130-fa501fef-6129-4766-8b46-1ab716925244.png)